src/test/spec/read_write_concern/document.rs

```diff
@@ -15,80 +12,108 @@ struct TestFile {
 
-    for (key, value) in write_concern_doc {
-        match (&key[..], value) {
-            ("w", Bson::Int32(i)) => {
-                write_concern.w = Some(Acknowledgment::from(i as u32));
-            }
-            ("w", Bson::String(s)) => {
-                write_concern.w = Some(Acknowledgment::from(s));
-            }
-            ("journal", Bson::Boolean(b)) => {
-                write_concern.journal = Some(b);
-            }
-            ("wtimeoutMS", Bson::Int32(i)) if i > 0 => {
-                write_concern.w_timeout = Some(Duration::from_millis(i as u64));
+        if let Some(specified_write_concern_document) = test_case.write_concern {
+            let specified_write_concern =
+                match bson::from_document::<WriteConcern>(specified_write_concern_document) {
+                    Ok(write_concern) => {
+                        let is_valid = write_concern.validate().is_ok();
+                        assert_eq!(
+                            is_valid, test_case.valid,
+                            "Write concern deserialization/validation should fail: {}",
+                            description
+                        );
+                        write_concern
+                    }
+                    Err(err) => {
+                        assert!(
+                            !test_case.valid,
+                            "Write concern deserialization should succeed but got {:?}: {}",
+                            description, err
+                        );
+                        continue;
+                    }
+                };
+
+            if let Some(is_server_default) = test_case.is_server_default {
+                assert_eq!(
+                    specified_write_concern.is_empty(),
+                    is_server_default,
+                    "is_server_default is {} but write concern is {:?}: {}",
+                    is_server_default,
+                    &specified_write_concern,
+                    description
+                );
             }
-            ("wtimeoutMS", Bson::Int32(_)) => {
-                // WriteConcern has an unsigned integer for the wtimeout field, so this is
-                // impossible to test.
-                return None;
+
+            if let Some(is_acknowledged) = test_case.is_acknowledged {
+                assert_eq!(
+                    specified_write_concern.is_acknowledged(),
+                    is_acknowledged,
+                    "is_acknowledged is {} but write concern is {:?}: {}",
+                    is_acknowledged,
+                    &specified_write_concern,
+                    description
+                );
             }
-            _ => {}
-        };
-    }
 
-    Some(write_concern)
-}
+            let actual_write_concern_document = bson::to_document(&specified_write_concern)
+                .unwrap_or_else(|_| {
+                    panic!(
+                        "Write concern serialization should succeed: {}",
```

Comment:
Might be good to include the error in the output here for diagnostics (likewise for the read concern doc).